### PR TITLE
Picture layers cache pictures as textures when their pictures don't change

### PR DESCRIPTION
**Pull request by chinmaygarde**

### CODE REVIEW

File: sky/compositor/clip_path_layer.cc [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+ClipPathLayer::ClipPathLayer() {
+}
+
+ClipPathLayer::~ClipPathLayer() {
+}
+
+void ClipPathLayer::set_clip_path(const SkPath& clip_path) {
+  clip_path_ = clip_path;
+}
```
Comment (abarth) [inline sky/compositor/clip_path_layer.cc]: Usually we leave trivial getters and setters like these in the header.

File: sky/compositor/clip_rect_layer.cc [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+namespace compositor {
+ClipRectLayer::ClipRectLayer() {
+}
+
+ClipRectLayer::~ClipRectLayer() {
+}
+
+void ClipRectLayer::set_clip_rect(const SkRect& clip_rect) {
+  clip_rect_ = clip_rect;
+}
```
Comment (abarth) [inline sky/compositor/clip_rect_layer.cc]: ditto

File: sky/compositor/color_filter_layer.cc [hunk 3]
```diff
@@ -0,0 +1,33 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "sky/compositor/color_filter_layer.h"
+
+namespace sky {
+namespace compositor {
```
Comment (abarth) [inline sky/compositor/color_filter_layer.cc]: We usually leave a blank line under the last namespace declaration.  "git cl format" should clean up this sort of thing for you.

File: sky/compositor/color_filter_layer.cc [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+void ColorFilterLayer::Paint(GrContext* context, SkCanvas* canvas) {
+  RefPtr<SkColorFilter> color_filter =
+      adoptRef(SkColorFilter::CreateModeFilter(color_, transfer_mode_));
+  SkPaint paint;
+  paint.setColorFilter(color_filter.get());
+  canvas->saveLayer(&paint_bounds(), &paint);
+  PaintChildren(context, canvas);
+  canvas->restore();
+}
+}  // compositor
```
Comment (abarth) [inline sky/compositor/color_filter_layer.cc]: We usually have a blank line above the first close-namespace }

File: sky/compositor/picture_layer.cc [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+                                        const SkRect& paintBounds) {
+  // Step 1: Create a texture from the context's texture provider
+
+  GrSurfaceDesc desc;
+  desc.fWidth = paintBounds.width();
+  desc.fHeight = paintBounds.height();
+  desc.fFlags = kRenderTarget_GrSurfaceFlag;
+  desc.fConfig = kRGBA_8888_GrPixelConfig;
+
+  auto texture = context->textureProvider()->createTexture(desc, true);
```
Comment (abarth) [inline sky/compositor/picture_layer.cc]: Can you put the type in here so we can see whether its a raw pointer or a smart pointer?

File: sky/compositor/picture_layer.cc [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+  DCHECK(picture_);
+
+  if (last_picture_id_ == picture_->uniqueID()) {
+    // The last picture painted was the same as this one. Cache this into an
+    // offscreen surface and render that instead
+    if (!cached_image_) {
+      // Generate the cached image
+      RefPtr<SkImage> image =
+          ImageFromPicture(context, picture_.get(), paint_bounds());
+      cached_image_.swap(image);
```
Comment (abarth) [inline sky/compositor/picture_layer.cc]: Why not just:

cached_image_ = ImageFromPicture(context, picture_.get(), paint_bounds());

?

File: sky/compositor/picture_layer.cc [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    cached_image_.release();
+  }
+
+  last_picture_id_ = picture_->uniqueID();
+
+  canvas->save();
+  canvas->translate(offset_.x(), offset_.y());
+
+  if (cached_image_) {
+    canvas->drawImage(cached_image_.get(), 0, 0);
```
Comment (abarth) [inline sky/compositor/picture_layer.cc]: In the drawImage case, we can avoid the save/translate/restore by putting offset_.x() and offset_.y() in for the 0 values here.

File: sky/compositor/picture_layer.cc (hunk 6 above)
Comment (chinmaygarde) [inline sky/compositor/picture_layer.cc]: Fixed

File: sky/compositor/picture_layer.cc [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+static RefPtr<SkImage> ImageFromPicture(GrContext* context,
+                                        SkPicture* picture,
+                                        const SkRect& paintBounds) {
+  // Step 1: Create a texture from the context's texture provider
+
+  GrSurfaceDesc desc;
+  desc.fWidth = paintBounds.width();
+  desc.fHeight = paintBounds.height();
```
Comment (abarth) [inline sky/compositor/picture_layer.cc]: Don't we need to scale the size of the texture we produce by the matrix?  Maybe you're planning to do that in a later patch.

Comment (chinmaygarde) [inline sky/compositor/picture_layer.cc]: Yeah, I spent too much time on the MPD path. Just wanted to submit some patches. Will fix

File: sky/compositor/picture_layer.cc (hunk 7 above)
Comment (chinmaygarde) [inline sky/compositor/picture_layer.cc]: Fixed

File: sky/compositor/picture_layer.cc [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+  RefPtr<SkImage> image = adoptRef(SkImage::NewFromTexture(
+      context, backendDesc, kPremul_SkAlphaType, &ImageReleaseProc, texture));
+  return image;
+}
+
+void PictureLayer::Paint(GrContext* context, SkCanvas* canvas) {
+  DCHECK(picture_);
+
+  if (last_picture_id_ == picture_->uniqueID()) {
```
Comment (abarth) [inline sky/compositor/picture_layer.cc]: When will this become true?  We get a new PictureLayer every frame...

**Comment by abarth** [comment]
LGTM

File: sky/compositor/picture_layer.cc (hunk 9 above)
Comment (chinmaygarde) [inline sky/compositor/picture_layer.cc]: You're right. I was under the impression that the PictureLayer would be the same but the picture itself may or may not change.

Comment (abarth) [inline sky/compositor/picture_layer.cc]: That's why I had the PictureTable in my draft patch.  That was a way of remembering state about SkPicture objects across layer trees.

**Pull request merged**